### PR TITLE
Improved probability !=0 normalization and moved to model-level

### PR DESCRIPTION
**Pull request by jwmueller**
Improved probability normalization and changed it to apply at the model-level across all models rather than in trainer.

Will prevent predicted-probabilities = 0 or exceeding [0,1] when activated.

Currently activated for all models if certain eval_metric is detected like log_loss that doesn't like 0s.

Will resolve issues like:  https://github.com/awslabs/autogluon/issues/473

### CODE REVIEW

**Comment by szha** [comment]
Job PR-481-1 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-481/1/index.html

**Comment by szha** [comment]
Job PR-481-2 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-481/2/index.html

File: docs/tutorials/tabular_prediction/tabular-quickstart.md [hunk 1]
```diff
@@ -107,4 +107,4 @@ predictor_age = task.fit(train_data=train_data, output_directory="agModels-predi
 performance = predictor_age.evaluate(test_data)
 ```
 
-Note that we didn't need to tell AutoGluon this is a regression problem, it automatically inferred this from the data and reported the appropriate performance metric (RMSE by default).
+Note that we didn't need to tell AutoGluon this is a regression problem, it automatically inferred this from the data and reported the appropriate performance metric (RMSE by default). To specify a particular evaluation metric other than the default, set the `eval_metric` argument of `fit()` and AutoGluon will tailor its models to optimize your metric (e.g. `eval_metric = mean_absolute_error`). For evaluation metrics where higher values are worse (like RMSE), AutoGluon may sometimes flips their sign and print them as negative values during training (as it internally assumes higher values are better).
```
Comment (Innixma) [inline docs/tutorials/tabular_prediction/tabular-quickstart.md:110]: mean_absolute_error -> 'mean_absolute_error'

File: autogluon/utils/tabular/ml/models/tabular_nn/tabular_nn_model.py [hunk 2]
```diff
@@ -96,6 +96,10 @@ def __init__(self, path: str, name: str, problem_type: str, objective_func, stop
         self._architecture_desc = None
         self.optimizer = None
         self.verbosity = None
+        if self.stopping_metric is not None and self.objective_func == roc_auc and self.stopping_metric == log_loss:
+            self.stopping_metric = roc_auc  # NN is overconfident so early stopping with logloss can halt training too quick
```
Comment (Innixma) [inline autogluon/utils/tabular/ml/models/tabular_nn/tabular_nn_model.py:100]: See my comment re-updating abstract_model predict_proba, I don't think this is necessary after that change. You would just have to compute normalize_pred_proba_stopping_metric during training and pass that into predict_proba.

File: autogluon/utils/tabular/ml/models/lgb/lgb_model.py [hunk 3]
```diff
@@ -125,26 +127,16 @@ def fit(self, X_train=None, Y_train=None, X_test=None, Y_test=None, dataset_trai
...
-        elif self.problem_type == MULTICLASS:
-            return y_pred_proba
-        else:
-            if len(y_pred_proba.shape) == 1:
-                return y_pred_proba
-            elif y_pred_proba.shape[1] > 2:  # Should this ever happen?
-                return y_pred_proba
-            else:  # Should this ever happen?
-                return y_pred_proba[:, 1]
+        if (self.problem_type == BINARY) and (len(y_pred_proba.shape) > 1) and (y_pred_proba.shape[1] > 1):
```
Comment (Innixma) [inline autogluon/utils/tabular/ml/models/lgb/lgb_model.py:132]: The previous code was quite verbose, but it was that way because it minimized the amount of checks required of the data, and thus I think is faster than the condensed version and recommend this change be reverted. Instead of return y_pred_proba, just set y_pred_proba=, then have the normalization code at the end as currently done.

File: autogluon/utils/tabular/ml/models/lgb/lgb_model.py [hunk 4]
```diff
@@ -125,26 +127,16 @@ def fit(self, X_train=None, Y_train=None, X_test=None, Y_test=None, dataset_trai
     def predict_proba(self, X, preprocess=True):
```
Comment (Innixma) [inline autogluon/utils/tabular/ml/models/lgb/lgb_model.py:127]: Add argument: `normalize_pred_probas=None`

```
if normalize_pred_probas is None:
    normalize_pred_probas = self. normalize_pred_probas
```

This enables us to toggle this on during training if stopping_metric requires normalization. We wouldn't have to change stopping_metric to roc_auc in NN.

Comment (Innixma) [reply autogluon/utils/tabular/ml/models/lgb/lgb_model.py:127]: Ideally we want to move this function to be _predict_proba, with predict_proba having the normalize_pred_probas arg to avoid duplicate code which is starting to pop up. predict_proba should only be defined in abstract_model, and individual models can implement _predict_proba if they have custom logic.

```
def predict_proba(self, X, preprocess=True, normalize=None):
    if normalize is None:
         normalize = self.normalize_pred_probas
    y_pred_proba = self._predict_proba(X=X, preprocess=preprocess)
    if normalize:
        y_pred_proba = normalize_pred_probas(y_pred_proba, self.problem_type)
    return y_pred_proba
```

With this, I don't think any changes to the model predict_proba functions are necessary.

**Comment by szha** [comment]
Job PR-481-3 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-481/3/index.html

**Comment by szha** [comment]
Job PR-481-4 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-481/4/index.html

File: autogluon/utils/tabular/ml/models/abstract/abstract_model.py [hunk 5]
```diff
@@ -184,21 +190,18 @@ def predict_proba(self, X, preprocess=True):
             X = self.preprocess(X)
 
         if self.problem_type == REGRESSION:
-            return self.model.predict(X)
-
-        y_pred_proba = self.model.predict_proba(X)
+            y_pred_proba = self.model.predict(X)
```
[the remark this answers is not in the document]
Comment (jwmueller) [reply autogluon/utils/tabular/ml/models/abstract/abstract_model.py]: done

File: autogluon/utils/tabular/ml/models/abstract/abstract_model.py [hunk 6]
```diff
@@ -77,6 +77,12 @@ def __init__(self, path: str, name: str, problem_type: str, objective_func, stop
         else:
             self.stopping_metric = stopping_metric
 
+        self.normalize_predprobs = False  # do probabilistic predictions need to be renormalized
```
[the remark this answers is not in the document]
Comment (jwmueller) [reply autogluon/utils/tabular/ml/models/abstract/abstract_model.py]: done

File: autogluon/utils/tabular/ml/utils.py [hunk 7]
```diff
@@ -168,3 +154,52 @@ def shuffle_df_rows(X: DataFrame, seed=0, reset_index=True):
...
+
+def normalize_pred_probas(y_predprob, problem_type, eps=1e-7):
+    """ Remaps the predicted probabilities to ensure there are no zeros (needed for certain metrics like log-loss)
+        and that no predicted probability exceeds [0,1] (eg. in distillation when classification is treated as regression).
+        Args:
+            y_predprob: 1D (for binary classification) or 2D (for multiclass) numpy array of predicted probabilities
+            problem_type: We only consider normalization if the problem_type is one of: [BINARY, MULTICLASS, SOFTCLASS]
+            eps: controls around how far from 0 remapped predicted probabilities should be (larger `eps` means predicted probabilities will lie further from 0).
+    """
+    if (problem_type == REGRESSION) and (len(y_predprob.shape) > 1) and (y_predprob.shape[1] > 1):
```
[the remark this answers is not in the document]
Comment (jwmueller) [reply autogluon/utils/tabular/ml/utils.py:167]: because of future distillation PR

File: autogluon/utils/tabular/ml/utils.py [hunk 8]
```diff
@@ -168,3 +154,52 @@ def shuffle_df_rows(X: DataFrame, seed=0, reset_index=True):
...
+        problem_type = SOFTCLASS  # this was MULTICLASS problem converted to REGRESSION (as done in distillation)
+
+    if problem_type in [BINARY, REGRESSION]:
```
[the remark this answers is not in the document]
Comment (jwmueller) [reply autogluon/utils/tabular/ml/utils.py:170]: because of future distillation PR

File: autogluon/utils/tabular/ml/utils.py [hunk 9]
```diff
@@ -168,3 +154,52 @@ def shuffle_df_rows(X: DataFrame, seed=0, reset_index=True):
...
+    else:
+        raise ValueError(f"Invalid problem_type")
+    return y_predprob
+
+
+def normalize_binary_probas(y_predprob, eps):
+    (pmin,pmax) = (eps, 1-eps)  # predicted probs outside this range will be remapped into (0,1)
+    which_toobig = y_predprob > pmax
+    if np.sum(which_toobig) > 0:  # remap overly large probs
+        y_predprob = np.logical_not(which_toobig)*y_predprob + which_toobig*(1-(eps*np.exp(-(y_predprob-pmax))))
```
[the remark this answers is not in the document]
Comment (jwmueller) [reply autogluon/utils/tabular/ml/utils.py:189]: done

File: autogluon/utils/tabular/ml/utils.py [hunk 10]
```diff
@@ -168,3 +154,52 @@ def shuffle_df_rows(X: DataFrame, seed=0, reset_index=True):
     if reset_index:
         X_shuffled.reset_index(inplace=True, drop=True)
     return X_shuffled
+
+
+def normalize_pred_probas(y_predprob, problem_type, eps=1e-7):
```
[the remark this answers is not in the document]
Comment (jwmueller) [reply autogluon/utils/tabular/ml/utils.py:159]: 1e-7 is safe value to use for now.  Proper value for most datasets is probably bigger, but that could lead to suboptimal performance for super easy classification problems (e.g. the trivial ones in OpenML)

File: autogluon/utils/tabular/ml/utils.py [hunk 11]
```diff
@@ -168,3 +154,52 @@ def shuffle_df_rows(X: DataFrame, seed=0, reset_index=True):
...
+        y_predprob = np.logical_not(which_toobig)*y_predprob + which_toobig*(1-(eps*np.exp(-(y_predprob-pmax))))
+    which_toosmall = y_predprob < pmin
+    if np.sum(which_toosmall) > 0:  # remap overly small probs
+        y_predprob = np.logical_not(which_toosmall)*y_predprob + which_toosmall*eps*np.exp(-(pmin-y_predprob))
+    return y_predprob
+
+
+def normalize_multi_probas(y_predprob, eps):
+    min_predprob = np.min(y_predprob)
+    if min_predprob < 0:  # ensure nonnegative rows
```
[the remark this answers is not in the document]
Comment (jwmueller) [reply autogluon/utils/tabular/ml/utils.py:199]: because of future distillation PR

File: docs/tutorials/tabular_prediction/tabular-quickstart.md (hunk 1 above)
Comment (jwmueller) [reply docs/tutorials/tabular_prediction/tabular-quickstart.md]: done

File: autogluon/utils/tabular/ml/models/tabular_nn/tabular_nn_model.py (hunk 2 above)
Comment (jwmueller) [reply autogluon/utils/tabular/ml/models/tabular_nn/tabular_nn_model.py:99]: This is not because of anything being broken.  This PR would've fixed the old NN issue without this change.  This change is because NN should not be early stopping based on logloss.  Logloss early stopping can make NN stop training too early, because NN tends to get overconfident very quickly (which is not a problem for AUC metric). Metrics like accuracy, AUC can still improve on validation set even after logloss has peaked.

File: autogluon/utils/tabular/ml/models/lgb/lgb_model.py (hunk 3 above)
Comment (jwmueller) [reply autogluon/utils/tabular/ml/models/lgb/lgb_model.py]: done

File: autogluon/utils/tabular/ml/models/lgb/lgb_model.py (hunk 4 above)
Comment (jwmueller) [reply autogluon/utils/tabular/ml/models/lgb/lgb_model.py]: done

**Comment by szha** [comment]
Job PR-481-5 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-481/5/index.html

File: autogluon/utils/tabular/ml/models/lr/lr_model.py [hunk 12]
```diff
@@ -140,7 +140,7 @@ def fit(self, X_train, Y_train, X_test=None, Y_test=None, time_limit=None, **kwa
 
         self.model = model.fit(X_train, Y_train)
 
-    def predict_proba(self, X, preprocess=True):
+    def _predict_proba(self, X, preprocess=True):
         X = self.preprocess(X, is_train=False, model_specific_preprocessing=True)
         return super().predict_proba(X, preprocess=False)
```
Comment (Innixma) [inline autogluon/utils/tabular/ml/models/lr/lr_model.py:145]: super()._predict_proba(), also make sure this is done everywhere else in predict_proba calls, otherwise normalization will be triggered twice.

**Comment by szha** [comment]
Job PR-481-6 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-481/6/index.html

File: autogluon/utils/tabular/ml/models/lr/lr_model.py (hunk 12 above)
Comment (jwmueller) [reply autogluon/utils/tabular/ml/models/lr/lr_model.py]: good catch, fixed it
